subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaCompileExecutableUtils.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile;
+
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.tasks.compile.ForkOptions;
+import org.gradle.api.tasks.compile.JavaCompile;
+import org.gradle.jvm.toolchain.JavaToolchainSpec;
+import org.gradle.jvm.toolchain.internal.SpecificInstallationToolchainSpec;
+
+import javax.annotation.Nullable;
+import java.io.File;
+
+public class JavaCompileExecutableUtils {
+
+    @Nullable
+    public static JavaToolchainSpec getExecutableOverrideToolchainSpec(JavaCompile task, ObjectFactory objectFactory) {
+        ForkOptions forkOptions = task.getOptions().getForkOptions();
+        File customJavaHome = forkOptions.getJavaHome();
```

Comment:
Does it make sense to take `fork = true/false` into account here?

Reply:
I wanted to postpone this change to the main PR, because it's effects are more noticable.

I looked at it again now, and realised that taking the `fork` into account here would actually be very good.
It will result into the `getCompilerTool()` respecting the `executable` and `javaHome` options (by ignoring them with `fork = false`) even before we go into the Java compile spec factory. This will have more effects:

- The https://github.com/gradle/gradle/issues/22398 will get fixed
- The `getJavaVersion()` input will change from the wrong one corresponding to the toolchain of the executable to the toolchain of the build JVM that would actually be running the task
- And `isToolchainCompatibleWithJava8()` will become less broken for the case of `fork = false`.

So, this would be somewhat a breaking change. But we wanted to fix this problems in the main PR anyways. So, I would add the `fork` option handling here in this PR.

Reply:
Sounds good!